File: 2023/20.py
import math

input = open(f"{__file__.replace('py', 'input')}").read().strip().split("\n")
input = [x.split(" -> ") for x in input]
types = {}
modules = {}
acc = {}
state = {}

for line in input:
    modules[line[0][1:]] = line[1].split(", ")
    types[line[0][1:]] = line[0][0]
    state[line[0][1:]] = False

for t in [x for x in types if types[x] == "&"]:
    acc[t] = {}
    for k in modules:
        for vv in modules[k]:
            if vv == t:
                acc[t][k] = False

low = 0
high = 0
t = 0
cache = {}
steps = []
rx = ["br", "fk", "lf", "rz"]

while True:
    queue = [["roadcaster", "", False]]
    while queue:
        to, pulse_from, pulse = queue.pop(0)

        if pulse:
            high += 1
        else:
            if to in rx and to in cache:
                steps.append(t - cache[to])
            cache[to] = t
            low += 1

        if len(steps) == len(rx):
            print(math.lcm(*steps))
            exit()

        if not to in modules:
            continue

        if types[to] == "b":
            for new_to in modules[to]:
                queue.append([new_to, to, pulse])
        elif types[to] == "%":
            if not pulse:
                state[to] = not state[to]
                pulse = state[to]
                for new_to in modules[to]:
                    queue.append([new_to, to, pulse])
        elif types[to] == "&":
            acc[to][pulse_from] = pulse
            pulse = not all(x for x in acc[to].values())
            for new_to in modules[to]:
                queue.append([new_to, to, pulse])

    t += 1
    if t == 1000:
        print(low * high)

File: 2023/20.rb
input = File.read(File.basename(__FILE__).gsub('rb', 'input')).strip.split("\n").map { _1.split(' -> ') }
types = input.map { [_1[0][1..], _1[0][0]] }.to_h
input = input.map { [_1[1..], _2.split(', ')] }.to_h

state = input.keys.map { [_1, false] }.to_h
acc = {}
types.select { _2 == '&' }.each_key do |t|
  acc[t] ||= {}
  input.each do |k, v|
    v.each do |from|
      acc[t][k] = false if from == t
    end
  end
end

low = 0
high = 0
t = 0
cache = {}
steps = []
rx = %w[br fk lf rz]
loop do
  queue = [['roadcaster', '', false]]
  until queue.empty?
    to, from, pulse = queue.shift

    if pulse
      high += 1
    else
      steps << t - cache[to] if rx.include?(to) && cache[to]
      cache[to] = t
      low += 1
    end

    if steps.size == rx.size
      p steps.reduce(1, :lcm)
      exit
    end

    next unless input[to]

    case types[to]
    when 'b'
      input[to].each do |new_to|
        queue << [new_to, to, pulse]
      end
    when '%'
      unless pulse
        state[to] = !state[to]
        pulse = state[to]
        input[to].each do |new_to|
          queue << [new_to, to, pulse]
        end
      end
    when '&'
      acc[to][from] = pulse
      pulse = !acc[to].values.all? { _1 == true }
      input[to].each do |new_to|
        queue << [new_to, to, pulse]
      end
    end
  end
  t += 1
  p low * high if t == 1000
end
